Add einops-style `rearrange` (#44)
stanford-crfm/haliax: docs/rearrange.md

`reshape`, `transpose`, and other shape-manipulation operations. Haliax provides a subset of its functionality
(specifically `rearrange` and not `repeat` or `reduce`, which are less useful in named code). The syntax has been generalized to named
tensors.

If you're used to einops, the syntax should be familiar, with the main differences being specifying names
and the additional "unordered" syntax for selecting dimensions by name.

!!! warning
    This syntax is fairly new. It is pretty well-tested, but it is possible that there are bugs.

### Examples

Examples are probably the best way to get a feel for the syntax:

```python
import haliax as hax
import jax.random as jrandom

N = hax.Axis("N", 32)
C = hax.Axis("C", 3)
H = hax.Axis("H", 64)
W = hax.Axis("W", 64)

x = hax.random.normal(jrandom.PRNGKey(0), (N, C, H, W))

# transpose/permute axes
y = hax.rearrange(x, "N C H W -> N H W C")
# names don't have to match with positional syntax
z = hax.rearrange(x, "num ch h w -> num h w ch")
# ellipsis can be used to specify the rest of the dimensions
z = hax.rearrange(x, "N C ... -> N ... C")

> # unordered syntax allows for select dimensions by name

Reviewer: ```suggestion
# `{ }` is used for unordered axes. It allows for selecting dimensions by name
```
I was initially confused by `{ }`, until I read the syntax section.

# unordered version
y = hax.rearrange(x, "{(H: patch_h H) (W: patch_w W) C } -> ... (P: patch_h patch_w) (C: C H W)", H=4, W=4)
```

### Bindings: Aliasing and Sizing

In the above examples we used keyword arguments to give sizes to split axes, which is the same
as in einops. However, we can also use bindings to alias axes. For example:

```python
# this produces the same result as the previous example
y2 = hax.rearrange(x, "N C (patch_h foo) (patch_w bar) -> N (P: patch_h patch_w) (C: C foo bar)", foo=hax.Axis("H", 4), bar=hax.Axis("W", 4))
assert y.axes == y2.axes
```

This example is a bit contrived, but it lets us use shorter or different names in the string which is occasionally useful.
You can actually just use strings to specify aliases, which is sometimes useful as well.
For instance, if we wanted "P" to actually be called "patch", but wanted to keep the short syntax, we could do:

```python
y3 = hax.rearrange(x, "N C (nh ph) (nw pw) -> N (P: nh nw) (C: C ph pw)", P="patch", pw=4, ph=4)
```

### Differences from einops

As you may have noticed, there are some differences from einops:

* Merged axes must have a name: `(C: C H W)` instead of `(C H W)`

> * The unordered syntax with `{ ... }` is new but hopefully not too confusing: select dimensions by name, and optionally split them using, e.g., `(N: step microbatch)`.

Reviewer: ```suggestion
* The unordered syntax with `{ ... }` is new but hopefully not too confusing
* Select dimensions by name, and optionally split them using `:` , e.g., `(N: step microbatch)`.
```

Let me know if I understand it incorrectly. This seems to be two different points.

Reviewer: The use of ellipsis doesn't seem confusing to me at least (maybe the confusion happens the code implementation rather than reading it). It seems to be a useful thing to have.

Author: i guess i meant that it's either `{ N }` or `{ (N: x y) }` (with the latter being splitting). I'll rework

* As discussed immediately above, you can use bindings to specify axis objects for names as well as sizes.

### Syntax

Semiformally, the syntax is an lhs and an rhs. The lhs is either ordered or unordered:

* *ordered lhs*: a sequence of axis variable names or (named or anonymous) split axes, separated by spaces or commas, and up to one ellipsis
* *unordered lhs*: a sequence of axis names or named split axes, surrounded by `{}`, separated by spaces or commas

> * An *axis variable name* is an identifier (that need not correspond to an axis name in the input or output.)

Reviewer: ```suggestion
* An *axis variable name* is an identifier (that does not need to correspond to an axis name in the input or output.)
```

Reviewer: I don't fully understand the difference between axis variable name vs. axis name here. 
In the example of `N = hax.Axis("N", 32)`, is `"N"` the axis name and `N` the axis variable name?

Author: added a note. that point is that the string in an axis variable is arbitrary, while an axis name must refer to an actual axis
